drivers/include/periph/freqm.h

```diff
@@ -0,0 +1,121 @@
+ * determine with a 24-bit accuracy the frequency of another clock.
+ *
+ *
+ * @{
+ *
+ * @file
+ * @brief       FREQM peripheral driver interface definitions
+ *
+ * @author      Urs Gompper <urs.gompper@ml-pa.com>
+ */
+
+#ifndef PERIPH_FREQM_H
+#define PERIPH_FREQM_H
+
+#include "periph_cpu.h"
+
+/* Channel Enable Mask */
+#define GCLK_PCHCTRL_CHEN_Msk   (_U_(0x1) << GCLK_PCHCTRL_CHEN_Pos)
+/* Enable Mask */
+#define FREQM_CTRLA_ENABLE_Msk  (_U_(0x1) << FREQM_CTRLA_ENABLE_Pos)
+/* Start Measurement Mask */
+#define FREQM_CTRLB_START_Msk   (_U_(0x1) << FREQM_CTRLB_START_Pos)
+/* Measurement Done Interrupt Enable Mask */
+#define FREQM_INTENSET_DONE_Msk (_U_(0x1) << FREQM_INTENSET_DONE_Pos)
+/* Measurement Done Mask */
+#define FREQM_INTFLAG_DONE_Msk  (_U_(0x1) << FREQM_INTFLAG_DONE_Pos)
+/* FREQM Status Mask */
+#define FREQM_STATUS_BUSY_Msk   (_U_(0x1) << FREQM_STATUS_BUSY_Pos)
+/* Sticky Count Value Overflow Mask */
+#define FREQM_STATUS_OVF_Msk    (_U_(0x1) << FREQM_STATUS_OVF_Pos)
+
+
+typedef void (*freqm_callback_t)(uintptr_t context);
+
+/**
+ * @brief   Frequency meter configuration
+ *
+ */
+struct freqm_config {
+    uint8_t msr_clk_src;    /* GCLK source select for measurement */
+    uint8_t ref_clk_src;    /* GCLK source select for reference */
+    uint8_t ref_clk_cnt;    /* Measurement duration in number of reference clock cycles */
+    uint32_t ref_clk_freq;  /* Frequency of reference clock */
+};
+
+/**
+ * @brief   Start frequency measurement
+ *
+ * @return                  true if measurement started successfully
+ * @return                  false if frequency measurement is already ongoing
+ */
+bool freqm_start(void);
+
+/**
+ * @brief   Check if an overflow condition occured
+ *
+ * In case of an overflow condition either the number of reference clock cycles
+ * must be reduced or a faster reference clock must be configured.
+ *
+ * @return                  0 if no overflow occured
+ * @return                  1 if an overflow occured
+ */
+int freqm_error_get(void);
+
+/**
+ * @brief   Read number of periods of measured clock and calculate frequency
+ *
+ * @param[in] config        configuration
+ *
+ * @return                  calculated frequency
+ */
+uint32_t freqm_frequency_get(struct freqm_config *const config);
+
+/**
+ * @brief   Query if a frequency measurement is ongoing
+ *
+ * @return                  true if frequency measurement is ongoing
+ * @return                  false if there's no ongoing frequency measurement
+ */
+bool freqm_is_busy(void);
+
+/**
+ * @brief Register a callback function
+ *
+ * @param freqmCallback     callback function
+ * @param context
+ */
+void freqm_callback_register(freqm_callback_t freqmCallback, uintptr_t context);
+
+/**
+ * @brief   Initialize the frequency meter
+ *
+ * @param[in] config        configuration
+ *
+ * @return                  0 on success
+ * @return                  -1 on invalid ADC line
```

Comment:
IMO, it would be better to return -EINVAL here. (and use standard errnos for this driver) instead. This way, the driver would be more consistent with other drivers

Reply:
Or if the config is expected to be compile-time const, you can also use `void` here and fail an `assert()` when something is wrong.

That way a wrong configuration is caught early in development.